src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,757 @@
+}
+
+uint32_t RoutingManager::GetPrefixExpireDelay(uint32_t aValidLifetime)
+{
+    uint32_t delay;
+
+    if (aValidLifetime * static_cast<uint64_t>(1000) > Timer::kMaxDelay)
+    {
+        delay = Timer::kMaxDelay;
+    }
+    else
+    {
+        delay = aValidLifetime * 1000;
+    }
+
+    return delay;
+}
+
+void RoutingManager::HandleRouterAdvertisement(const Ip6::Address &aSrcAddress,
+                                               const uint8_t *     aBuffer,
+                                               uint16_t            aBufferLength)
+{
+    OT_UNUSED_VARIABLE(aSrcAddress);
+
+    using RouterAdv::Option;
+    using RouterAdv::PrefixInfoOption;
+    using RouterAdv::RouterAdvMessage;
+
+    bool           needReevaluate = false;
+    const uint8_t *optionsBegin;
+    uint16_t       optionsLength;
+    const Option * option;
+
+    VerifyOrExit(aBufferLength >= sizeof(RouterAdvMessage));
+
+    otLogInfoBr("received Router Advertisement from %s on interface %s", aSrcAddress.ToString().AsCString(),
+                GetInfraIfName());
+
+    optionsBegin  = aBuffer + sizeof(RouterAdvMessage);
+    optionsLength = aBufferLength - sizeof(RouterAdvMessage);
+
+    option = Option::GetNextOption(nullptr, optionsBegin, optionsLength);
+    while (option != nullptr)
+    {
+        if (option->GetType() == Option::Type::kPrefixInfo)
+        {
+            auto &      pio = *static_cast<const PrefixInfoOption *>(option);
+            Ip6::Prefix prefix;
+
+            prefix.Set(pio.GetPrefix(), pio.GetPrefixLength());
+            if (IsValidOnLinkPrefix(prefix))
+            {
+                if (pio.GetValidLifetime() == 0 && mDiscoveredOnLinkPrefix == prefix)
+                {
+                    otLogInfoBr("invalidate discovered on-link prefix %s", prefix.ToString().AsCString());
+                    mDiscoveredOnLinkPrefixInvalidTimer.Stop();
+                    mDiscoveredOnLinkPrefix.Clear();
+                    needReevaluate = true;
+                }
+                else
+                {
+                    otLogInfoBr("set discovered on-link prefix to %s, valid lifetime: %u seconds",
+                                prefix.ToString().AsCString(), pio.GetValidLifetime());
+
+                    // We keep tracking the latest on-link prefix.
+                    mDiscoveredOnLinkPrefixInvalidTimer.Start(GetPrefixExpireDelay(pio.GetValidLifetime()));
+                    mDiscoveredOnLinkPrefix = prefix;
+
+                    // Stop Router Solicitation if we found a valid on-link prefix.
+                    // Otherwise, we wait till the Router Solicitation process timeouted.
+                    // So the maximum delay before the Border Router starts advertising
+                    // its own on-link prefix is 9 (4 + 4 + 1) seconds.
+                    mRouterSolicitTimer.Stop();
+                    needReevaluate = true;
+                }
+            }
```

Comment:
What happens in the code if the received RA contains multiple Prefix Info options, all being valid OnLinkPrefix and all having non-zero lifetime?

If I follow the code correctly, sine valid lifetime is non-zero, the `if (pio.GetValidLifetime() == 0 && mDiscoveredOnLinkPrefix == prefix)` would not be true, so we go to the `else` block and then keep updating the `mDiscoveredOnLinkPrefix`. It would basically set to the last Prefix Option. Is this what we want?

Reply:
Yes, the behavior is intended. We tracks the fact that there is some valid on-link prefix, but not all on-link prefixes. The discovered on-link prefix is to stop ourselves from sending my local on-link prefix, so I think as long as there is a valid on-link prefix, we don't need to send my on-link prefix.